README.md
Description
===========

This cookbook installs and configures a Zope ZEO-Server.

Requirements
============

Cookbooks
---------

- python


Attributes
==========

See `attributes/default.rb` for default values.

General attributes
------------------

* `node[:zeoserver][:user]` - User the Zope ZEO-Server will run as. Defaults
  to `zope`.
* `node[:zeoserver][:dir]` - Location for the ZEO-Server installation. Defaults
  to `/srv/zeoserver`.
* `node[:zeoserver][:dir_log]` - Location for the log files. Defaults to
  `/var/log/zeoserver`.
* `node[:zeoserver][:dir_var]` - Used to configure the base directory for all
  things going into var. Defaults to `/srv/zeoserver/var`.
* `node[:zeoserver][:dir_filestorage]` - Directory for database files. Defaults
  to `/srv/zeoserver/var/filestorage`.
* `node[:zeoserver][:dir_blobstorage]` - Directory for blob files. Defaults to
  `/srv/zeoserver/var/blobstorage`.
* `node[:zeoserver][:install_method]` - Method to install the ZEO-Server with.
  Defaults to `buildout`.

Process related attributes
--------------------------

* `node[:zeoserver][:zeo_address]` - The port for the ZEO server (either
  specify the port number only (with `127.0.0.1` as default) or you use the
  format `host:port`). Defaults to `127.0.0.1:8100`.
* `node[:zeoserver][:monitor_address]` - The address at which the monitor
  server should listen. The monitor server provides server statistics in a
  simple text format. Defaults to `127.0.0.1:8200`.
* `node[:zeoserver][:effective_user]` - The name of the effective user for the
  ZEO process. This causes the process to run under the user account the
  process has been started with. Defaults to the `node[:zeoserver][:user]`.
* `node[:zeoserver][:socket_name]` - The filename where ZEO will write its
  socket file. Defaults to `zeo.zdsock`.

Packing related attributes
--------------------------

* `node[:zeoserver][:do_pack]` - Enable weekly scheduled packing of the ZODB.
  Defaults to `true`.
* `node[:zeoserver][:pack_days]` - How many days of history should the zeopack
  script retain. Defaults to `1`.
* `node[:zeoserver][:pack_gc]` - Can be set to false to disable garbage
  collection as part of the pack. Defaults to `true`.
* `node[:zeoserver][:pack_keep_old]` - Can be set to false to disable the
  creation of *.fs.old files before the pack is run. Defaults to `true`.
* `node[:zeoserver][:pack_user]` - If the ZEO server uses authentication, this
  is the username used by the zeopack script to connect to the ZEO server. The
  default is no user.
* `node[:zeoserver][:pack_password]` - If the ZEO server uses authentication,
  this is the password used by the zeopack script to connect to the ZEO server.
  The default is no password.
* `node[:zeoserver][:pack_minute]` - Minute at which the pack should be done.
  Only in combination with `do_pack`. Defaults to `5`.
* `node[:zeoserver][:pack_hour]` - Hour at which the pack should be done. Only
  in combination with `do_pack`. Defaults to `0`.
* `node[:zeoserver][:pack_weekday]` - Weekday at which the pack should be done.
  Only in combination with `do_pack`. Defaults to `0` (sunday).

Backup related attributes
-------------------------

* `node[:zeoserver][:do_backup]` - Enable backup and do a daily one via cron.
  Defaults to `true`.
* `node[:zeoserver][:dir_backups]` - Directory for database and blob backups.
  Defaults to `/srv/zeoserver/var/backups`.
* `node[:zeoserver][:backup_minute]` - Minute at which the backup should be
  done. Only in combination with `do_backup`. Defaults to `5`.
* `node[:zeoserver][:backup_hour]` - Hour at which the backup should be done.
  Only in combination with `do_backup`. Defaults to `1`.

Buildout
--------

* `node[:zeoserver][:virtualenv]` - Directory for the virtualenv. Defaults to
  `/srv/zeoserver/venv`.


Usage
=====

Include the recipe on your node or role that fits how you wish to install the
Zope ZEO-Server on your system per the recipes section above. Modify the
attributes as required in your role to change how various configuration is
applied per the attributes section above. In general, override attributes in
the role should be used when changing attributes.

default
-------

Installs a Zope ZEO-Server by `buildout`, `pip` or `package` depending on the
chosen install method. The ZEO-Server is registered as a service and started
automatically.

Installs daily backups and weekly database packing via cron jobs (can be
disabled - see attributes).

buildout
--------

Installs a ZEO-Server via buildout.

pip
---

Installs a ZEO-Server via pip.

package
-------

Installs a ZEO-Server from packages (coming soon).


License and Author
==================

Author:: Thomas Massmann <thomas.massmann@inqbus.de>

Copyright:: 2012, Inqbus

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# CHANGELOG for zeoserver

This file is used to list changes made in each version of zeoserver.

## 0.1.2 (unreleased):


## 0.1.1 (2012-11-14):

* Installs ZEO-server via pip (default is still buildout).


## 0.1.0 (2012-11-13):

* Initial release of zeoserver.
* Installs ZEO-Server via buildout.
* Adds zeoserver init script.
* Creates cron job for daily backups and weekly database packs.

- - - 
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/pip.rb
#
# Cookbook Name:: zeoserver
# Recipe:: pip
#
# Copyright 2012, Inqbus
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Install dependencies, prepare common directories.
include_recipe "zeoserver::commons"


python_virtualenv node[:zeoserver][:virtualenv] do
  owner node[:zeoserver][:user]
  action :create
end

cookbook_file "#{node[:zeoserver][:dir]}/requirements.txt" do
  source "requirements.txt"
  owner node[:zeoserver][:user]
  mode 0644
  notifies :run, "execute[requirements_update]", :immediately
end

execute "requirements_update" do
  cwd node[:zeoserver][:dir]
  command "#{node[:zeoserver][:virtualenv]}/bin/pip install -U -r #{node[:zeoserver][:dir]}/requirements.txt"
  user node[:zeoserver][:user]
  action :nothing
end

python_pip "requirements" do
  package_name "#{node[:zeoserver][:dir]}/requirements.txt"
  virtualenv node[:zeoserver][:virtualenv]
  options "-r"
  action :install
end

directory "#{node[:zeoserver][:dir]}/bin" do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

link "#{node[:zeoserver][:dir]}/bin/zeopack" do
  to "#{node[:zeoserver][:virtualenv]}/bin/zeopack"
  owner node[:zeoserver][:user]
end

template "zeoserver" do
  path "#{node[:zeoserver][:dir]}/bin/zeoserver"
  source "zeoserver.erb"
  owner node[:zeoserver][:user]
  mode "0755"
end

directory "#{node[:zeoserver][:dir]}/parts/zeoserver" do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

directory "#{node[:zeoserver][:dir]}/parts/zeoserver/bin" do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

directory "#{node[:zeoserver][:dir]}/parts/zeoserver/etc" do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

template "zeo.conf" do
  path "#{node[:zeoserver][:dir]}/parts/zeoserver/etc/zeo.conf"
  source "zeo.conf.erb"
  owner node[:zeoserver][:user]
  mode "0644"
  # notifies :restart, resources(:service => "zeoserver")
end

template "runzeo" do
  path "#{node[:zeoserver][:dir]}/parts/zeoserver/bin/runzeo"
  source "runzeo.erb"
  owner node[:zeoserver][:user]
  mode "0755"
end

link "#{node[:zeoserver][:dir]}/parts/zeoserver/bin/zeoctl" do
  to "#{node[:zeoserver][:virtualenv]}/bin/zeoctl"
  owner node[:zeoserver][:user]
end

%w{backups blobstoragebackups snapshotbackups blobstoragesnapshots}.each do |dir|
  directory "#{node[:zeoserver][:dir_backups]}/#{dir}" do
    mode "0755"
    owner node[:zeoserver][:user]
    recursive true
    action :create
  end
end

template "backup" do
  path "#{node[:zeoserver][:dir]}/bin/backup"
  source "backup.erb"
  owner node[:zeoserver][:user]
  mode "0755"
end

template "restore" do
  path "#{node[:zeoserver][:dir]}/bin/restore"
  source "restore.erb"
  owner node[:zeoserver][:user]
  mode "0755"
end

template "snapshotbackup" do
  path "#{node[:zeoserver][:dir]}/bin/snapshotbackup"
  source "snapshotbackup.erb"
  owner node[:zeoserver][:user]
  mode "0755"
end

template "snapshotrestore" do
  path "#{node[:zeoserver][:dir]}/bin/snapshotrestore"
  source "snapshotrestore.erb"
  owner node[:zeoserver][:user]
  mode "0755"
end

# Install and register services.
include_recipe "zeoserver::services"


recipes/commons.rb
#
# Cookbook Name:: zeoserver
# Recipe:: commons
#
# Author:: Thomas Massmann (<thomas.massmann@inqbus.de>)
#
# Copyright 2012, Inqbus
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Python is required.
include_recipe "python::default"

# Install rsync
package "rsync"

service "zeoserver" do
  provider Chef::Provider::Service::Init::Debian
  supports :restart => true, :start => true, :stop => true
end

user node[:zeoserver][:user] do
  shell "/bin/sh"
  home node[:zeoserver][:dir]
  system true
  action :create
end

directory node[:zeoserver][:dir] do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

directory node[:zeoserver][:dir_log] do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

directory node[:zeoserver][:dir_var] do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

directory node[:zeoserver][:dir_backups] do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

directory node[:zeoserver][:dir_filestorage] do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end

directory node[:zeoserver][:dir_blobstorage] do
  mode "0755"
  owner node[:zeoserver][:user]
  recursive true
  action :create
end


recipes/buildout.rb
#
# Cookbook Name:: zeoserver
# Recipe:: buildout
#
# Author:: Thomas Massmann (<thomas.massmann@inqbus.de>)
#
# Copyright 2012, Inqbus
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Install dependencies, prepare common directories.
include_recipe "zeoserver::commons"


directory "#{node[:zeoserver][:dir]}/downloads" do
  mode "0755"
  owner node[:zeoserver][:user]
  action :create
end

directory "#{node[:zeoserver][:dir]}/extends-cache" do
  mode "0755"
  owner node[:zeoserver][:user]
  action :create
end

python_virtualenv node[:zeoserver][:virtualenv] do
  owner node[:zeoserver][:user]
  action :create
end

cookbook_file "#{node[:zeoserver][:dir]}/bootstrap.py" do
  source "bootstrap.py"
  owner node[:zeoserver][:user]
  mode 0644
end

template "#{node[:zeoserver][:dir]}/buildout.cfg" do
  source "buildout.cfg.erb"
  owner node[:zeoserver][:user]
  mode 0644
  variables({
    :dir_filestorage => node[:zeoserver][:dir_filestorage],
    :dir_blobstorage => node[:zeoserver][:dir_blobstorage],
    :dir_backups => node[:zeoserver][:dir_backups],
    :dir_log => node[:zeoserver][:dir_log],
    :dir_var => node[:zeoserver][:dir_var],
    :zeo_address => node[:zeoserver][:zeo_address],
    :monitor_address => node[:zeoserver][:monitor_address],
    :effective_user => node[:zeoserver][:effective_user],
    :socket_name => node[:zeoserver][:socket_name],
    :pack_days => node[:zeoserver][:pack_days],
    :pack_gc => node[:zeoserver][:pack_gc],
    :pack_keep_old => node[:zeoserver][:pack_keep_old],
    :pack_user => node[:zeoserver][:pack_user],
    :pack_password => node[:zeoserver][:pack_password],
  })
  notifies :run, "execute[buildout]", :immediately
  notifies :restart, resources(:service => "zeoserver")
end

execute "buildout" do
  cwd node[:zeoserver][:dir]
  command "#{node[:zeoserver][:virtualenv]}/bin/python bootstrap.py && ./bin/buildout"
  user node[:zeoserver][:user]
  action :nothing
end

# Install and register services.
include_recipe "zeoserver::services"


recipes/services.rb
#
# Cookbook Name:: zeoserver
# Recipe:: commons
#
# Author:: Thomas Massmann (<thomas.massmann@inqbus.de>)
#
# Copyright 2012, Inqbus
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

template "zeoserver_init" do
  path "/etc/init.d/zeoserver"
  source "zeoserver_init.erb"
  owner "root"
  group "root"
  mode "0755"
  notifies :restart, resources(:service => "zeoserver")
end

service "zeoserver" do
  action [:enable, :start]
end

cron "zeoserver-backup" do
  minute node[:zeoserver][:backup_minute]
  hour node[:zeoserver][:backup_hour]
  command "#{node[:zeoserver][:dir]}/bin/backup -q"
  user node[:zeoserver][:user]
  if node[:zeoserver][:do_backup]
    action :create
  else
    action :delete
  end
end

cron "zeoserver-pack" do
  minute node[:zeoserver][:pack_minute]
  hour node[:zeoserver][:pack_hour]
  weekday node[:zeoserver][:pack_weekday]
  command "#{node[:zeoserver][:dir]}/bin/zeopack #{node[:zeoserver][:zeo_address]}"
  user node[:zeoserver][:user]
  if node[:zeoserver][:do_pack]
    action :create
  else
    action :delete
  end
end


metadata.rb
maintainer        "Inqbus"
maintainer_email  "thomas.massmann@inqbus.de"
license           "apache2"
description       "Installs and configures a Zope ZEO-Server."
long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version           "0.1.2"

depends           "python"


recipe "zeoserver", "Installs a ZEO-Server based on the chosen install method. Registers and starts a service. Installs cron jobs."
recipe "zeoserver::buildout", "Installs a ZEO-Server via buildout."
recipe "zeoserver::pip", "Installs a ZEO-Server via pip."


# General Attributes.
attribute "zeoserver/user",
  :display_name => "Zope User",
  :description => "User the Zope ZEO-Server will run as.",
  :default => "zope"

attribute "zeoserver/dir",
  :display_name => "ZEO-Server Directory",
  :description => "Location for the ZEO-Server installation.",
  :default => "/srv/zeoserver"

attribute "zeoserver/dir_log",
  :display_name => "Log Directory",
  :description => "Location for the log files.",
  :default => "/var/log/zeoserver"

attribute "zeoserver/dir_var",
  :display_name => "Var Directory",
  :description => "Used to configure the base directory for all things going into var.",
  :default => "/srv/zeoserver/var"

attribute "zeoserver/dir_filestorage",
  :display_name => "Database Directory",
  :description => "Directory for database files.",
  :default => "/srv/zeoserver/var/filestorage"

attribute "zeoserver/dir_blobstorage",
  :display_name => "BLOB Directory",
  :description => "Directory for blob files.",
  :default => "/srv/zeoserver/var/blobstorage"

attribute "zeoserver/install_method",
  :display_name => "Installation Method",
  :description => "How the ZEO-Server should be installed.",
  :default => "buildout"


# Process related attributes
attribute "zeoserver/zeo_address",
  :display_name => "zeo_address",
  :description => "Give a port for the ZEO server (either specify the port number only (with '127.0.0.1' as default) or you use the format host:port).",
  :default => "127.0.0.1:8100"

attribute "zeoserver/monitor_address",
  :display_name => "monitor_address",
  :description => "The address at which the monitor server should listen. The monitor server provides server statistics in a simple text format.",
  :default => "127.0.0.1:8200"

attribute "zeoserver/effective_user",
  :display_name => "effective_user",
  :description => "The name of the effective user for the ZEO process. This causes the process to run under the user account the process has been started with.",
  :default => "zope"

attribute "zeoserver/socket_name",
  :display_name => "socket_name",
  :description => "The filename where ZEO will write its socket file. Defaults to zeo.zdsock.",
  :default => "zeo.zdsock"


# Packing related attributes
attribute "zeoserver/do_pack",
  :display_name => "Enable Database Packing",
  :description => "Enable weekly scheduled packing of the ZODB.",
  :default => "true"

attribute "zeoserver/pack_days",
  :display_name => "pack_days",
  :description => "How many days of history should the zeopack script retain.",
  :default => "1"

attribute "zeoserver/pack_gc",
  :display_name => "pack_gc",
  :description => "Can be set to false to disable garbage collection as part of the pack.",
  :default => "true"

attribute "zeoserver/pack_keep_old",
  :display_name => "pack_keep_old",
  :description => "Can be set to false to disable the creation of *.fs.old files before the pack is run.",
  :default => "true"

attribute "zeoserver/pack_user",
  :display_name => "pack_user",
  :description => "If the ZEO server uses authentication, this is the username used by the zeopack script to connect to the ZEO server.",
  :default => nil

attribute "zeoserver/pack_password",
  :display_name => "pack_password",
  :description => "If the ZEO server uses authentication, this is the password used by the zeopack script to connect to the ZEO server.",
  :default => nil

attribute "zeoserver/pack_minute",
  :display_name => "pack_minute",
  :description => "Minute at which the pack should be done. Only in combination with 'do_pack'.",
  :default => "5"

attribute "zeoserver/pack_hour",
  :display_name => "pack_hour",
  :description => "Hour at which the pack should be done. Only in combination with 'do_pack'.",
  :default => "0"

attribute "zeoserver/pack_weekday",
  :display_name => "pack_weekday",
  :description => "Weekday at which the pack should be done. Only in combination with 'do_pack'.",
  :default => "0"


# Backup related attributes
attribute "zeoserver/do_backup",
  :display_name => "Enable Backup",
  :description => "Enable daily scheduled backups.",
  :default => "true"

attribute "zeoserver/dir_backups",
  :display_name => "Backups Directory",
  :description => "Directory for database and blob backups.",
  :default => "/srv/zeoserver/var/backups"

attribute "zeoserver/backup_minute",
  :display_name => "backup_minute",
  :description => "Minute at which the backup should be done. Only in combination with 'do_backup'.",
  :default => "5"

attribute "zeoserver/backup_hour",
  :display_name => "backup_hour",
  :description => "Hour at which the backup should be done. Only in combination with 'do_backup'.",
  :default => "1"


# Buildout related attributes
attribute "zeoserver/virtualenv",
  :display_name => "Virtualenv Directory",
  :description => "Directory for the virtualenv. Only used for buildout.",
  :default => "/srv/zeoserver/venv"
